### PR TITLE
[Core] Install SkyPilot runtime in separate env

### LINKED ISSUE #2722: [core] Ray version change breaks SkyPilot cluster
If the user's setup installs a newer version of ray than the one running on SkyPilot remote cluster, SkyPilot will get stuck at streaming logs. Many packages (e.g., `vllm==0.2.0`) tend to install a newer version of ray. 

SkyPilot should:
1. Protect the ray installation/use a custom env for the core cluster operation
2. Raise a better error is the remote ray cluster malfunctions.

## Minimal repro:
```yaml
setup: |
  pip install "ray==2.7.1"
run: |
  echo hi
```

```console
$ sky launch task.yaml

...

I 10-18 21:37:59 cloud_vm_ray_backend.py:3358] Job submitted with Job ID: 1
I 10-19 04:38:01 log_lib.py:431] Start streaming logs for job 1.
<Stuck>
```

## Workaround
Install all dependencies in a new conda env. 

## Debug
ssh into cluster and run `ray status`:
```console
(base) gcpuser@ray-test-2ea4-head-60220eef-compute:~$ ray status
Traceback (most recent call last):
  File "python/ray/_raylet.pyx", line 2929, in ray._raylet.check_health
  File "python/ray/_raylet.pyx", line 457, in ray._raylet.check_status
ray.exceptions.RaySystemError: System error: Ray cluster at 10.128.0.7:6380 has version 2.4.0, but this processis running Ray version 2.7.1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/ray", line 8, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.10/site-packages/ray/scripts/scripts.py", line 2498, in main
    return cli()
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/ray/scripts/scripts.py", line 1974, in status
    if not ray._raylet.check_health(address):
  File "python/ray/_raylet.pyx", line 2935, in ray._raylet.check_health
RuntimeError: System error: Ray cluster at 10.128.0.7:6380 has version 2.4.0, but this processis running Ray version 2.7.1.
```

### PR DESCRIPTION
**Pull request by Michaelvll**
Blocked by #3572
Moved #2801 here due to the merging issues and our latest support for explicitly python / ray path.

**This PR enables:**
1. Users to **install dependencies in the base environment** without worrying about conflicting with SkyPilot runtime.
2. Users to **have an image with python 3.12 as the default python version** (which is not supported by SkyPilot runtime).
3. A more SkyPilot native and less verbose axolotl example with the support of the latest axolotl docker image.

Closes #2722
Mitigate #2673

## Background
Installing every skypilot runtime in the base environment can cause issues when users also tries to install their dependencies in the base environment on the remote VM. It is very easy to cause the VM to become in an unexpected state. We now move all the skypilot runtime to a new conda environment for better isolation from the users' own Python environment.

### New cluster

**TL;DR: there is a negligible negative effect to the launching time for all the clouds with the separate env.**

<details>

<summary>Results on AWS, GCP, Kubernetes</summary>

#### AWS
`multitime -n 5 sky launch -y --cloud aws --cpus 2`
master
```
            Mean        Std.Dev.    Min         Median      Max
real        152.313     7.229       142.269     152.518     164.088     
user        6.756       0.527       6.405       6.498       7.800       
sys         0.721       0.116       0.562       0.714       0.923  
```
This PR
<details>

```
            Mean        Std.Dev.    Min         Median      Max
real        154.497     2.993       151.134     154.496     157.933     
user        6.841       0.419       6.516       6.665       7.659       
sys         0.712       0.079       0.637       0.684       0.853
```

</details>

```
            Mean        Std.Dev.    Min         Median      Max
real        149.659     6.103       143.091     148.129     160.476     
user        6.454       0.064       6.374       6.439       6.559       
sys         0.695       0.029       0.645       0.699       0.727  
```

#### GCP 
`multitime -n 5 sky launch -y --cloud gcp --cpus 2`
master
```
            Mean        Std.Dev.    Min         Median      Max
real        129.164     5.770       121.261     129.621     135.799     
user        17.263      0.141       17.135      17.191      17.499      
sys         1.764       0.075       1.675       1.730       1.863   
```
This PR 
<details>

(21s slower, due to the installation of dependencies for ray and skypilot)
```
            Mean        Std.Dev.    Min         Median      Max
real        150.368     4.356       144.007     150.592     156.529     
user        17.784      0.586       17.283      17.671      18.895      
sys         1.849       0.128       1.702       1.834       2.064    
```

</details>

after adding a6f6996
```
            Mean        Std.Dev.    Min         Median      Max
real        124.762     4.736       120.339     122.328     133.610     
user        17.562      0.585       17.147      17.307      18.724      
sys         1.875       0.107       1.727       1.867       2.032  
```

#### Kubernetes

master:
```
            Mean        Std.Dev.    Min         Median      Max
real        54.512      1.991       51.994      54.134      56.943      
user        4.770       0.611       4.416       4.489       5.988       
sys         0.633       0.064       0.558       0.645       0.734 
```

This PR:
```
            Mean        Std.Dev.    Min         Median      Max
real        56.873      3.327       53.555      54.552      61.738      
user        4.729       0.469       4.490       4.494       5.668       
sys         0.525       0.074       0.465       0.485       0.665
```

</details>

### Launch on existing cluster
`sky launch -c existing-cluster --cpus 2 --cloud kubernetes -y`
`multitime -n 5 sky launch -c existing-cluster echo hi`

Master:
```
            Mean        Std.Dev.    Min         Median      Max
real        49.564      1.980       47.926      48.969      53.362      
user        3.318       0.446       3.035       3.139       4.204       
sys         0.423       0.104       0.340       0.376       0.627   
```
This PR:
```
            Mean        Std.Dev.    Min         Median      Max
real        46.420      0.826       45.826      46.087      48.059      
user        3.096       0.046       3.043       3.080       3.169       
sys         0.409       0.021       0.395       0.399       0.451    
```

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch -c test-ax --cloud kubernetes  axolotl.yaml --env HF_TOKEN`
  - [x] `sky launch -c test-ax --cloud gcp -i 0 --down --use-spot axolotl.yaml --env HF_TOKEN`
  - [x] `sky launch -c test-ax --cloud gcp -i 0 --down --use-spot axolotl.yaml --env HF_TOKEN --image-id docker:winglian/axolotl:main-latest` (with python 3.12 in the docker image) 
  - [x] `sky launch --image-id ami-0df2a11dd1fe1f8e3 --cpus 2 --cloud aws --region us-east-1 -y -i20 --down conda env list` (AMI used in #2673)
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [x] All smoke tests: `pytest tests/test_smoke.py --aws`
- [x] All smoke tests: `pytest tests/test_smoke.py --kubernetes` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

**Comment by dongreenberg** [comment]
Cc @carolineechen

File: sky/skylet/attempt_skylet.py [hunk 1]
```diff
@@ -21,6 +21,8 @@ def restart_skylet():
         shell=True,
         check=False)
     subprocess.run(
+        # Activate python environment first to make sure skylet can find the
+        # cloud SDK for autostopping.
         f'nohup {constants.SKY_PYTHON_CMD} -m sky.skylet.skylet'
```
Comment (cblmemo) [inline sky/skylet/attempt_skylet.py:26]: this does not activate the python env?

File: sky/skylet/constants.py [hunk 2]
```diff
@@ -91,20 +101,29 @@
 # AWS's Deep Learning AMI's default conda environment.
 CONDA_INSTALLATION_COMMANDS = (
     'which conda > /dev/null 2>&1 || '
-    '{ wget -nc https://repo.anaconda.com/miniconda/Miniconda3-py310_23.11.0-2-Linux-x86_64.sh -O Miniconda3-Linux-x86_64.sh && '  # pylint: disable=line-too-long
+    '{ curl https://repo.anaconda.com/miniconda/Miniconda3-py310_23.11.0-2-Linux-x86_64.sh -o Miniconda3-Linux-x86_64.sh && '  # pylint: disable=line-too-long
     'bash Miniconda3-Linux-x86_64.sh -b && '
     'eval "$(~/miniconda3/bin/conda shell.bash hook)" && conda init && '
     'conda config --set auto_activate_base true && '
     # Use $(echo ~) instead of ~ to avoid the error "no such file or directory".
     # Also, not using $HOME to avoid the error HOME variable not set.
```
Comment (cblmemo) [inline sky/skylet/constants.py:109]: remove those comment as well?

File: tests/test_smoke.py [hunk 3]
```diff
@@ -1026,8 +1026,10 @@ def test_kubernetes_storage_mounts():
         'docker:nvidia/cuda:11.8.0-devel-ubuntu18.04',
         'docker:ubuntu:18.04',
         # Test latest image with python 3.11 installed by default.
```
Comment (cblmemo) [inline tests/test_smoke.py:1028]:
```suggestion
        # Test image with python 3.11 installed by default.
```

**Review by cblmemo — APPROVED** [review]
Thanks for adding this @Michaelvll ! It looks great to me :)) Left some nits

File: llm/axolotl/axolotl.yaml
Comment (concretevitamin) [inline llm/axolotl/axolotl.yaml:1]: We should update Axolotl's GitHub readme with this, maybe after 0.6.

File: llm/axolotl/axolotl-docker.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    accelerate launch -m axolotl.cli.train /sky_workdir/qlora.yaml
+
+envs:
+  HF_TOKEN: # TODO: Fill with your own huggingface token, or use --env to pass.
+  
+  
+
+
+  
+
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply llm/axolotl/axolotl-docker.yaml:35]: Done. Thanks!
